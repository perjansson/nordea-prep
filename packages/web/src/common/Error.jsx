import React from 'react'

import Message from 'common/Message'

const Error = ({ children }) => <Message>{children}</Message>

export default Error
