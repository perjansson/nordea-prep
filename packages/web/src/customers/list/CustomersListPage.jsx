import React from 'react'
import { Link } from 'react-router-dom'

import {
    Page,
    Header,
    FadeInOut,
    Title,
    BackLink,
    Footer,
} from '@nordea-prep/components'

import CustomersListContainer from 'customers/list/CustomersListContainer'

const CustomersListPage = () => (
    <Page>
        <Header>
            <FadeInOut>
                <Title className="title" emoji="ğŸ¤©">
                    customers list page
                </Title>
            </FadeInOut>
        </Header>
        <CustomersListContainer />
        <Footer>
            <Link to="/">
                <BackLink className="backLink">start page</BackLink>
            </Link>
        </Footer>
    </Page>
)

export default CustomersListPage
