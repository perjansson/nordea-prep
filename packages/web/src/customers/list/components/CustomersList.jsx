import React from 'react'
import { instanceOf } from 'prop-types'
import styled, { css } from 'styled-components'

import {
    List,
    ListHeaderRow,
    ListHeader,
    ListItems,
    ListValue,
} from '@nordea-prep/components/dist/List'

const gridTemplateColumnMixin = css`
    grid-template-columns: 1fr 1fr 1fr;
`

const fontMixin = css`
    color: white;
    text-shadow: 0 0 10px black, 0 0 5px darkblue;
    line-height: 56px;
`

const CustomerListHeaderRow = styled(ListHeaderRow)`
    ${gridTemplateColumnMixin};
    ${fontMixin};
    border-bottom: 2px dashed white;
`

const CustomerListItemRow = styled(ListItems)`
    ${gridTemplateColumnMixin};
    ${fontMixin};
`

const Wrapper = styled.div`
    width: 100vw;
    padding: 0 20px;
`

const CustomersList = ({ customers }) => (
    <Wrapper>
        <List
            items={customers}
            renderHeaderRow={() => (
                <CustomerListHeaderRow>
                    <ListHeader>Name</ListHeader>
                    <ListHeader>Phone</ListHeader>
                    <ListHeader>Email</ListHeader>
                </CustomerListHeaderRow>
            )}
            renderRow={({ name, phone, email }) => (
                <CustomerListItemRow key={`${name}_${phone}_${email}`}>
                    <ListValue>{name}</ListValue>
                    <ListValue>{phone}</ListValue>
                    <ListValue>{email}</ListValue>
                </CustomerListItemRow>
            )}
            data-test-name="customer-list"
        />
    </Wrapper>
)

CustomersList.propTypes = {
    customers: instanceOf(Array),
}

CustomersList.defaultProps = {
    customers: [],
}

export default CustomersList
