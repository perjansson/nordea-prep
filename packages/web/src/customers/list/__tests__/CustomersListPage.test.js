import React from 'react'

import {
    renderApp,
    mockDataFetch,
    mockDataFetchError,
    restoreFetchMock,
    flushAllPromises,
} from '__test__/util'

import CustomersListPage from 'customers/list/CustomersListPage'
import CustomersList from 'customers/list/components/CustomersList'
import Error from 'common/Error'

describe('<CustomersListPage />', () => {
    describe('integration test', () => {
        afterEach(restoreFetchMock)

        it('fetches and shows customers on load', async () => {
            const customersResponse = [
                {
                    id: 0,
                    name: 'Mina Watsica',
                    phone: '139-120-3133',
                    email: 'Gracie_Weber@hotmail.com',
                },
            ]
            mockDataFetch('customers', customersResponse)

            const { wrapper, store } = renderApp(<CustomersListPage />)

            // No rendered customers list initially
            expect(wrapper.find(CustomersList).exists()).toBeFalsy()

            await flushAllPromises()
            wrapper.update()

            // Customers list renderer after data fetch is complete
            const customersList = wrapper.find(CustomersList)
            expect(customersList.exists()).toBeTruthy()
            expect(customersList.prop('customers')).toEqual(customersResponse)
            expect(store.getState().customers.list).toEqual(customersResponse)
        })

        it('fetches and shows error message on fetch error', async () => {
            const errorMessage = 'specific error message'
            mockDataFetchError('customers', errorMessage)

            const { wrapper } = renderApp(<CustomersListPage />)
            await flushAllPromises()
            wrapper.update()

            const error = wrapper.find(Error)
            expect(error.exists()).toBeTruthy()
            expect(error.props().message).toContain(errorMessage)
        })
    })
})
