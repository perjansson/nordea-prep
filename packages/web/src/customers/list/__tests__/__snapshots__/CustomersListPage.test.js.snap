// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomersListPage /> integration test displays empty page while loading customers and full list when loaded 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MemoryRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <CustomersListPage>
        <styled.div>
          <div
            className="sc-bdVaJa eKvdSL"
          >
            <Component>
              <header>
                <styled.div>
                  <div
                    className="sc-bwzfXH egnLMR"
                  >
                    <Title
                      className="title"
                      emoji="ðŸ¤©"
                      onClick={[Function]}
                      style={Object {}}
                    >
                      <styled.h1
                        className="title"
                        onClick={[Function]}
                        style={Object {}}
                      >
                        <h1
                          className="title sc-htpNat hOBrmH"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          customers list page
                          <span
                            aria-label="smiley-with-star-eyes"
                            role="img"
                          >
                             
                            ðŸ¤©
                          </span>
                        </h1>
                      </styled.h1>
                    </Title>
                  </div>
                </styled.div>
              </header>
            </Component>
            <Connect(CustomersListContainer)>
              <CustomersListContainer
                customers={Array []}
                dispatch={[Function]}
                error={null}
                showCustomers={false}
              >
                <Error>
                  <Message>
                    <div
                      className="Message-klYsfs ktFSZq"
                    />
                  </Message>
                </Error>
              </CustomersListContainer>
            </Connect(CustomersListContainer)>
            <Component>
              <footer>
                <Link
                  replace={false}
                  to="/"
                >
                  <a
                    href="/"
                    onClick={[Function]}
                  >
                    <BackLink
                      className="backLink"
                    >
                      <div
                        className="backLink"
                      >
                        <styled.span>
                          <span
                            className="sc-ifAKCX brUUVN"
                          >
                            &lt;&lt; 
                          </span>
                        </styled.span>
                        <styled.span>
                          <span
                            className="sc-bxivhb keKaPJ"
                          >
                            to 
                            start page
                          </span>
                        </styled.span>
                        <styled.span>
                          <span
                            className="sc-ifAKCX brUUVN"
                          >
                             &lt;&lt;
                          </span>
                        </styled.span>
                      </div>
                    </BackLink>
                  </a>
                </Link>
              </footer>
            </Component>
          </div>
        </styled.div>
      </CustomersListPage>
    </Router>
  </MemoryRouter>
</Provider>
`;

exports[`<CustomersListPage /> integration test displays empty page while loading customers and full list when loaded 2`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MemoryRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <CustomersListPage>
        <styled.div>
          <div
            className="sc-bdVaJa eKvdSL"
          >
            <Component>
              <header>
                <styled.div>
                  <div
                    className="sc-bwzfXH egnLMR"
                  >
                    <Title
                      className="title"
                      emoji="ðŸ¤©"
                      onClick={[Function]}
                      style={Object {}}
                    >
                      <styled.h1
                        className="title"
                        onClick={[Function]}
                        style={Object {}}
                      >
                        <h1
                          className="title sc-htpNat hOBrmH"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          customers list page
                          <span
                            aria-label="smiley-with-star-eyes"
                            role="img"
                          >
                             
                            ðŸ¤©
                          </span>
                        </h1>
                      </styled.h1>
                    </Title>
                  </div>
                </styled.div>
              </header>
            </Component>
            <Connect(CustomersListContainer)>
              <CustomersListContainer
                customers={
                  Array [
                    Object {
                      "email": "Brooklyn77@yahoo.com",
                      "id": "1",
                      "name": "Marlene Ledner",
                      "phone": "482-056-9445",
                    },
                  ]
                }
                dispatch={[Function]}
                error={null}
                showCustomers={true}
              >
                <CustomersList
                  customers={
                    Array [
                      Object {
                        "email": "Brooklyn77@yahoo.com",
                        "id": "1",
                        "name": "Marlene Ledner",
                        "phone": "482-056-9445",
                      },
                    ]
                  }
                >
                  <Wrapper>
                    <div
                      className="CustomersList__Wrapper-kOJqXM gECnhh"
                    >
                      <List
                        data-test-name="customer-list"
                        items={
                          Array [
                            Object {
                              "email": "Brooklyn77@yahoo.com",
                              "id": "1",
                              "name": "Marlene Ledner",
                              "phone": "482-056-9445",
                            },
                          ]
                        }
                        renderHeaderRow={[Function]}
                        renderRow={[Function]}
                      >
                        <ListWrapper
                          data-test-name="customer-list"
                        >
                          <div
                            className="sc-EHOje IabPF"
                            data-test-name="customer-list"
                          >
                            <CustomerListHeaderRow>
                              <HeaderRow
                                className="CustomersList__CustomerListHeaderRow-pEiVw fbICpq"
                              >
                                <div
                                  className="CustomersList__CustomerListHeaderRow-pEiVw fbICpq sc-bZQynM fsLozJ"
                                >
                                  <Header>
                                    <div
                                      className="sc-gzVnrw cbxTUt"
                                    >
                                      Name
                                    </div>
                                  </Header>
                                  <Header>
                                    <div
                                      className="sc-gzVnrw cbxTUt"
                                    >
                                      Phone
                                    </div>
                                  </Header>
                                  <Header>
                                    <div
                                      className="sc-gzVnrw cbxTUt"
                                    >
                                      Email
                                    </div>
                                  </Header>
                                </div>
                              </HeaderRow>
                            </CustomerListHeaderRow>
                            <CustomerListItemRow
                              key="Marlene Ledner_482-056-9445_Brooklyn77@yahoo.com"
                            >
                              <Items
                                className="CustomersList__CustomerListItemRow-jteZpo BwHYO"
                              >
                                <div
                                  className="CustomersList__CustomerListItemRow-jteZpo BwHYO sc-htoDjs kUVcju"
                                >
                                  <Value>
                                    <div
                                      className="sc-dnqmqq iqeVoq"
                                    >
                                      Marlene Ledner
                                    </div>
                                  </Value>
                                  <Value>
                                    <div
                                      className="sc-dnqmqq iqeVoq"
                                    >
                                      482-056-9445
                                    </div>
                                  </Value>
                                  <Value>
                                    <div
                                      className="sc-dnqmqq iqeVoq"
                                    >
                                      Brooklyn77@yahoo.com
                                    </div>
                                  </Value>
                                </div>
                              </Items>
                            </CustomerListItemRow>
                          </div>
                        </ListWrapper>
                      </List>
                    </div>
                  </Wrapper>
                </CustomersList>
              </CustomersListContainer>
            </Connect(CustomersListContainer)>
            <Component>
              <footer>
                <Link
                  replace={false}
                  to="/"
                >
                  <a
                    href="/"
                    onClick={[Function]}
                  >
                    <BackLink
                      className="backLink"
                    >
                      <div
                        className="backLink"
                      >
                        <styled.span>
                          <span
                            className="sc-ifAKCX brUUVN"
                          >
                            &lt;&lt; 
                          </span>
                        </styled.span>
                        <styled.span>
                          <span
                            className="sc-bxivhb keKaPJ"
                          >
                            to 
                            start page
                          </span>
                        </styled.span>
                        <styled.span>
                          <span
                            className="sc-ifAKCX brUUVN"
                          >
                             &lt;&lt;
                          </span>
                        </styled.span>
                      </div>
                    </BackLink>
                  </a>
                </Link>
              </footer>
            </Component>
          </div>
        </styled.div>
      </CustomersListPage>
    </Router>
  </MemoryRouter>
</Provider>
`;
